### ISSUE TITLE
Anvil: Janky mouse movement

### ISSUE THREAD

**Issue by maroider** [Smithay/smithay#224]
When runnig Anvil with `cargo run --features logind -- --tty-udev`, the mouse cursor acts as if my mouse has horribly inconsistent tracking. I've already tried running Weston on the same system, and it didn't experience the same issue (although Weston had a hard time displaying anything on the drm backend, so kudos to Anvil on that one). Sway has none of the above issues.

**Comment by kchibisov** [comment]
I'd recommend to try `--release` version, but I'm not sure that it'll help you here?

**Comment by maroider** [comment]
Yeah, I've tried that already but I forgot to mention it. :/

**Comment by kchibisov** [comment]
What GPU are you using?

**Comment by PolyMeilex** [comment]
Same here, just letting know.

**Comment by maroider** [comment]
I'm using an NVIDIA GeForce GTX 970 with the nouveau drivers (through mesa 20.1.2-1 on Arch).

**Comment by kchibisov** [comment]
One thing that you can test is proprietary drivers with anvil. You'll need to do  `modprobe nvidia-drm modeset=1` though.

**Comment by Drakulix** [comment]
Some log files would also be quite helpful to see what anvil selects to run on your system. anvil is using hardware cursors by default (with no real fallback right now) and this might be a case, where we would want to disable them in the future.

**Comment by maroider** [comment]
`modprobe` didn't work, but I've gone ahead and added `nvidia-drm.modeset=1` as a kernel parameter in  `/etc/default/grub`. I'll get back to you shortly with the results.

**Comment by maroider** [comment]
Using the proprietary driver yielded the same result.

@Drakulix Should I set a particular log level, or should I just set it to `TRACE`?

**Comment by Drakulix** [comment]
trace is likely to verbose. lets start with debug.

**Comment by maroider** [comment]
Where exactly do I set the log level? I'm not familiar with `slog` at all.

**Comment by Drakulix** [comment]
We really should start documenting this somewhere...
DEBUG is default for debug builds, while INFO is for release builds.
To change this, you want to compile with the [corresponding slog feature](https://github.com/slog-rs/slog/blob/master/Cargo.toml).
E.g. `cargo run --release --features "slog/release_max_level_debug" -- --tty-udev`

**Comment by maroider** [comment]
```
   Compiling slog v2.5.2
   Compiling anvil v0.0.1 (/home/markus/source/github.com/Smithay/smithay/anvil)
   Compiling slog-scope v4.3.0
   Compiling slog-term v2.6.0
   Compiling slog-async v2.5.0
   Compiling slog-stdlog v4.0.0
   Compiling smithay v0.2.0 (/home/markus/source/github.com/Smithay/smithay)
    Finished release [optimized] target(s) in 39.57s
     Running `/home/markus/source/github.com/Smithay/smithay/target/release/anvil --tty-udev`
Jul 02 13:17:58.286 INFO Starting anvil on a tty using udev
Jul 02 13:17:58.286 INFO Listening on wayland socket, name: wayland-0
Jul 02 13:17:58.286 INFO Trying to create logind session, session_type: auto, smithay_module: backend_session_auto
unable to lock lockfile /run/user/1000/wayland-0.lock, maybe another compositor is running
Jul 02 13:17:58.290 INFO Listening on wayland socket, name: wayland-1
Jul 02 13:17:58.290 INFO Initializing a xkbcommon handler with keymap query, options: None, variant: , layout: , model: , rules: , smithay_module: xkbcommon_handler, seat_name: seat0, smithay_module: seat_handler
Jul 02 13:17:58.293 INFO Loaded Keymap, name: Norwegian, smithay_module: xkbcommon_handler, seat_name: seat0, smithay_module: seat_handler
Jul 02 13:17:58.400 INFO Initializing a libinput backend, smithay_module: backend_libinput
Jul 02 13:17:58.401 INFO Trying to initialize AtomicDrmDevice, smithay_module: backend_drm_fallback
Jul 02 13:17:58.401 INFO AtomicDrmDevice initializing, smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:58.401 WARN Unable to become drm master, assuming unprivileged mode, smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:58.401 INFO Dropping device: Some("/dev/dri/card0"), smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:58.401 ERRO Failed to restore previous state. Error: unknown system error: EOPNOTSUPP: Operation not supported on transport endpoint, smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:58.401 WARN Failed to initialize preferred AtomicDrmDevice: DRM access error: Error enabling AtomicModesetting on device `Some("/dev/dri/card0")` (unknown system error: EOPNOTSUPP: Operation not supported on transport endpoint), smithay_module: backend_drm_fallback
Jul 02 13:17:58.401 INFO Falling back to fallback LegacyDrmDevice, smithay_module: backend_drm_fallback
Jul 02 13:17:58.401 INFO LegacyDrmDevice initializing, smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:58.401 WARN Unable to become drm master, assuming unprivileged mode, smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:58.888 INFO Drm device driver: Some("nouveau"), smithay_module: backend_drm_fallback
Jul 02 13:17:58.888 DEBG Creating gbm device, smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:17:58.924 DEBG Creating egl context from device, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:58.929 DEBG EGL No-Display Extensions: ["EGL_EXT_platform_base", "EGL_EXT_device_base", "EGL_EXT_device_enumeration", "EGL_EXT_device_query", "EGL_KHR_client_get_all_proc_addresses", "EGL_EXT_client_extensions", "EGL_KHR_debug", "EGL_KHR_platform_x11", "EGL_EXT_platform_x11", "EGL_EXT_platform_device", "EGL_KHR_platform_wayland", "EGL_EXT_platform_wayland", "EGL_MESA_platform_gbm", "EGL_KHR_platform_gbm", "EGL_MESA_platform_surfaceless"], smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:58.930 INFO EGL Initialized, smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:58.930 INFO EGL Version: (1, 5), smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:58.930 INFO EGL Extensions: ["EGL_ANDROID_blob_cache", "EGL_EXT_buffer_age", "EGL_EXT_create_context_robustness", "EGL_EXT_image_dma_buf_import", "EGL_EXT_image_dma_buf_import_modifiers", "EGL_KHR_cl_event2", "EGL_KHR_config_attribs", "EGL_KHR_create_context", "EGL_KHR_create_context_no_error", "EGL_KHR_fence_sync", "EGL_KHR_get_all_proc_addresses", "EGL_KHR_gl_colorspace", "EGL_KHR_gl_renderbuffer_image", "EGL_KHR_gl_texture_2D_image", "EGL_KHR_gl_texture_3D_image", "EGL_KHR_gl_texture_cubemap_image", "EGL_KHR_image", "EGL_KHR_image_base", "EGL_KHR_image_pixmap", "EGL_KHR_no_config_context", "EGL_KHR_reusable_sync", "EGL_KHR_surfaceless_context", "EGL_EXT_pixel_format_float", "EGL_KHR_wait_sync", "EGL_MESA_configless_context", "EGL_MESA_drm_image", "EGL_MESA_image_dma_buf_export", "EGL_MESA_query_driver", "EGL_WL_bind_wayland_display", ""], smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:58.930 INFO Initializing EGL Hardware Acceleration via "/dev/dri/card0"
Jul 02 13:17:58.964 INFO Connected: DVII
Jul 02 13:17:59.032 INFO Connected: HDMIA
Jul 02 13:17:59.138 INFO Initializing EglSurface, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.138 DEBG Trying to initialize EGL with OpenGLES 3.0, smithay_module: renderer_egl, smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.138 INFO Selected color format: PixelFormat { hardware_accelerated: true, color_bits: 24, alpha_bits: 8, depth_bits: 24, stencil_bits: 8, stereoscopy: false, multisampling: None, srgb: false }, smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.139 INFO EGL context created, smithay_module: renderer_egl, smithay_module: renderer_egl, smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.139 INFO Initializing GbmSurface, smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.139 INFO Initializing drm surface with mode Mode { name: "1920x1200", clock: 154000, size: (1920, 1200), hsync: (1968, 2000, 2080), vsync: (1203, 1209, 1235), hskew: 0, vscan: 0, vrefresh: 60 } and connectors [connector::Handle(73)], crtc: crtc::Handle(47), smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.345 DEBG (Re-)Initializing surface (with mode: 1920:1200), crtc: crtc::Handle(47), smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.345 DEBG EGL surface successfully created, smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.352 INFO Creating new wl_output, name: DVII, smithay_module: output_handler
Jul 02 13:17:59.352 INFO Initializing EglSurface, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.352 DEBG Trying to initialize EGL with OpenGLES 3.0, smithay_module: renderer_egl, smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.352 INFO Selected color format: PixelFormat { hardware_accelerated: true, color_bits: 24, alpha_bits: 8, depth_bits: 24, stencil_bits: 8, stereoscopy: false, multisampling: None, srgb: false }, smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.352 INFO EGL context created, smithay_module: renderer_egl, smithay_module: renderer_egl, smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.352 INFO Initializing GbmSurface, smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.352 INFO Initializing drm surface with mode Mode { name: "1920x1200", clock: 154000, size: (1920, 1200), hsync: (1968, 2000, 2080), vsync: (1203, 1209, 1235), hskew: 0, vscan: 0, vrefresh: 60 } and connectors [connector::Handle(78)], crtc: crtc::Handle(55), smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.558 DEBG (Re-)Initializing surface (with mode: 1920:1200), crtc: crtc::Handle(55), smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.558 DEBG EGL surface successfully created, smithay_module: renderer_egl, smithay_module: backend_egl, smithay_module: backend_drm_fallback
Jul 02 13:17:59.561 INFO Creating new wl_output, name: HDMIA, smithay_module: output_handler
Jul 02 13:17:59.561 DEBG Commiting new state, crtc: crtc::Handle(47), smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.594 INFO Adding connector: DVII, crtc: crtc::Handle(47), smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.628 INFO Setting new mode: "1920x1200", crtc: crtc::Handle(47), smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.628 DEBG Setting screen, crtc: crtc::Handle(47), smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.648 DEBG Commiting new state, crtc: crtc::Handle(55), smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.713 INFO Adding connector: HDMIA, crtc: crtc::Handle(55), smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.778 INFO Setting new mode: "1920x1200", crtc: crtc::Handle(55), smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.778 DEBG Setting screen, crtc: crtc::Handle(55), smithay_module: backend_drm, smithay_module: backend_drm_fallback
Jul 02 13:17:59.798 DEBG Importing cursor, crtc: crtc::Handle(47), smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:18:02.260 INFO New device "event1" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event0" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event14" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event15" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event16" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event17" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event18" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event19" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event20" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event10" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event2" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event3" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event6" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event9" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event7" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event8" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event22" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event23" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event24" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event25" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event26" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event27" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event28" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event29" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:02.260 INFO New device "event13" on seat "default", smithay_module: backend_libinput
Jul 02 13:18:14.110 DEBG Importing cursor, crtc: crtc::Handle(55), smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:18:15.411 DEBG Importing cursor, crtc: crtc::Handle(55), smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:18:17.146 DEBG Importing cursor, crtc: crtc::Handle(55), smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:18:18.213 DEBG Importing cursor, crtc: crtc::Handle(55), smithay_module: backend_gbm, smithay_module: backend_drm_fallback
Jul 02 13:18:19.487 DEBG key, state: Pressed, keycode: 29
Jul 02 13:18:19.487 DEBG keysym, keysym: Control_L, mods: ModifiersState { ctrl: true, alt: false, shift: false, caps_lock: false, logo: false, num_lock: false }, state: Pressed
Jul 02 13:18:19.612 DEBG key, state: Pressed, keycode: 56
Jul 02 13:18:19.612 DEBG keysym, keysym: Alt_L, mods: ModifiersState { ctrl: true, alt: true, shift: false, caps_lock: false, logo: false, num_lock: false }, state: Pressed
Jul 02 13:18:19.683 DEBG key, state: Pressed, keycode: 14
Jul 02 13:18:19.683 DEBG keysym, keysym: BackSpace, mods: ModifiersState { ctrl: true, alt: true, shift: false, caps_lock: false, logo: false, num_lock: false }, state: Pressed
Jul 02 13:18:19.683 INFO Quitting.
Jul 02 13:18:19.683 INFO Closing logind session, vt: Some(2), seat: seat0, id: 7, session_type: logind, smithay_module: backend_session, session_type: auto, smithay_module: backend_session_auto
```

**Comment by Drakulix** [comment]
Interesting, can you try if running with `SMITHAY_USE_LEGACY=1` makes the mouse behave better?

**Comment by maroider** [comment]
Still janky.

**Comment by Drakulix** [comment]
> Still janky.

If you hover over a window with a custom cursor (e.g. on weston-terminal) we switch to using a software cursor.
Does this help the perceived jankiness? I think this problem comes down to support full software and hardware cursors, while letting the user control, which is used, e.g. through another environment variable or by using a driver heuristic.

**Comment by maroider** [comment]
So, uh, how exactly do I launch an application while inside Anvil?

**Comment by Drakulix** [comment]
Mod+Enter should spawn `weston-terminal` if installed.

**Comment by vberger** [comment]
Or you can change the shortcuts here if you need:

https://github.com/Smithay/smithay/blob/b75d6524ce04005ba58a760f5bb90629676dfb5f/anvil/src/input_handler.rs#L290-L308

**Comment by maroider** [comment]
Hovering the cursor `weston-terminal`'s window had no perceivable effect on the jankiness.

**Comment by PolyMeilex** [comment]
Oh, so it's a different issue than mine, I thought that I have the same one, but for me, the software cursor works perfectly fine, only hardware cursor is "janky"

**Comment by Drakulix** [comment]
It seems that issues is related to both #227 and our use of hardware- vs. software-cursors.

**Comment by PolyMeilex** [comment]
> Oh, so it's a different issue than mine, I thought that I have the same one, but for me, the software cursor works perfectly fine, only hardware cursor is "janky"
> EDIT: #227 did not help.

On current master, it's no longer the case. It's been a while, so it's hard to pinpoint what fixed it, but it works as expected for me now too.

It was also fixed by #227 for ` LiHRaM` and `maroider`, so I believe the issue is fixed, and the issue probably™ can be closed.
